src/index.tsx

```diff
@@ -1,280 +1,142 @@
-    if (event.type === DispatchType.Update) {
-      // Internal action to update context
-      const nextContext = event.updater(state.context);
-      if (config.verbose) log('Context update', ['Previous Context', state.context], ['Next Context', nextContext]);
-      return {
-        ...state,
-        context: nextContext,
-      };
-    } else {
-      const currentState = config.states[state.value];
-
-      // Events can have a shortcut string format. We want the full object notation here
-      const eventObject = (typeof event.next === 'string' ? { type: event.next } : event.next) as Event<
-        Events,
-        EventString
-      >;
-
-      const nextState: Transition<Context, Events, State, EventString> | undefined =
-        currentState.on?.[eventObject.type] ?? config.on?.[eventObject.type];
 
-      // If there is no defined next state, return early
-      if (!nextState) {
-        if (config.verbose)
-          log(
-            `Current state doesn't listen to event type "${eventObject.type}".`,
-            ['Current State', state],
-            ['Event', eventObject]
-          );
-        return state;
+    if (internalEvent.type === "SEND") {
+      let sendable = internalEvent.sendable;
+      let event = typeof sendable === "string" ? { type: sendable } : sendable
+      let context = machineState.context;
+      let stateNode = R.get(definition.states, machineState.value)!;
+      let resolvedTransition =
+        R.get(R.fromMaybe(stateNode.on), event.type) ??
+        R.get(R.fromMaybe(definition.on), event.type)
+
+
+      if (!resolvedTransition) {
+        log(
+          `Current state doesn"t listen to event type "${event.type}".`,
+          ["Current State", machineState],
+          ["Event", event]
+        );
+        return machineState;
       }
 
-      let target: State;
-      if (typeof nextState === 'string') {
-        target = nextState;
-      } else {
-        target = nextState.target;
-        // If there are guards, invoke them and return early if the transition is denied
-        if (nextState.guard && !nextState.guard({ context: state.context, event: eventObject })) {
-          if (config.verbose)
-            log(
-              `Transition from "${state.value}" to "${target}" denied by guard`,
-              ['Event', eventObject],
-              ['Context', state.context]
-            );
-          return state;
-        }
+      let nextStateValue =
+        typeof resolvedTransition === "string" ? resolvedTransition :
+        resolvedTransition.guard === undefined ? resolvedTransition.target :
+        resolvedTransition.guard({ context, event }) ? resolvedTransition.target :
+        undefined;
+
+      if (nextStateValue === undefined) {
+        log(
+          `Current state doesn"t listen to event type "${event.type}".`,
+          ["Current State", machineState],
+          ["Event", event]
+        )
+        return machineState;
+      }
+      
```

Comment:
```suggestion
      log(`Transition from "${machineState.value}" to "${nextStateValue}"`, ['Event', eventObject])
      
```

nitpick: maybe rename `nextStateValue` to `nextMachineStateValue` I value consistency and specificity way higher than length of the a variable

Reply:
Agreed